fsspec/implementations/arrow.py

```diff
@@ -61,6 +61,20 @@ def protocol(self):
     def fsid(self):
         return "hdfs_" + tokenize(self.fs.host, self.fs.port)
 
+    @classmethod
+    def from_fs(cls, fs, **kwargs):
+        override = {"local": "file"}
+        type_name = override.get(fs.type_name, fs.type_name)
+
+        try:
+            fs.from_uri(f"{type_name}:///")
+            root_marker = "/"
+        except BaseException:
+            root_marker = ""
+
+        wrapper = type(cls.__name__, (cls,), {"root_marker": root_marker})
```

Comment:
So this is effectively a metaclass?

Reply:
There's no metaclasses involved, but it works similarly yes. For such a simple case, metaclasses might be overkill, but maybe a top-level function that isn't attached to `ArrowFSWrapper` instead? 

```python
def wrap_arrow_fs(fs: pyarrow.fs.FileSystem, **kwargs) -> ArrowFSWrapper:
    # build class dynamically based on fs
```

Because it's a class attribute, I don't think we can get around creating different classes per arrow filesystem. An alternative to the dynamic class construction would be to have wrapper subclasses for each `root_marker` variant and a lookup table (dict) based on the provided arrow filesystem.